File: app/unit-tests/src/org/commcare/utils/FirebaseMessagingUtilTest.java
package org.commcare.utils;

import org.junit.Test;

import java.io.IOException;
import java.util.concurrent.ExecutionException;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class FirebaseMessagingUtilTest {

    @Test
    public void testHasErrorInChain_directMatch() {
        Throwable error = new IOException("FIS_AUTH_ERROR");
        assertTrue(FirebaseMessagingUtil.hasErrorInChain(error, "FIS_AUTH_ERROR"));
    }

    @Test
    public void testHasErrorInChain_wrappedInExecutionException() {
        IOException cause = new IOException("FIS_AUTH_ERROR");
        ExecutionException wrapped = new ExecutionException(cause);
        assertTrue(FirebaseMessagingUtil.hasErrorInChain(wrapped, "FIS_AUTH_ERROR"));
    }

    @Test
    public void testHasErrorInChain_serviceNotAvailable() {
        IOException cause = new IOException("SERVICE_NOT_AVAILABLE");
        ExecutionException wrapped = new ExecutionException(cause);
        assertTrue(FirebaseMessagingUtil.hasErrorInChain(wrapped, "SERVICE_NOT_AVAILABLE"));
    }

    @Test
    public void testHasErrorInChain_noMatch() {
        Throwable error = new IOException("SOME_OTHER_ERROR");
        assertFalse(FirebaseMessagingUtil.hasErrorInChain(error, "FIS_AUTH_ERROR"));
    }

    @Test
    public void testHasErrorInChain_nullThrowable() {
        assertFalse(FirebaseMessagingUtil.hasErrorInChain(null, "FIS_AUTH_ERROR"));
    }

    @Test
    public void testHasErrorInChain_nullMessage() {
        Throwable error = new Throwable((String)null);
        assertFalse(FirebaseMessagingUtil.hasErrorInChain(error, "FIS_AUTH_ERROR"));
    }

    @Test
    public void testHasErrorInChain_selfReferencingCause() {
        SelfReferencingThrowable error = new SelfReferencingThrowable("FIS_AUTH_ERROR");
        // Should find the match on the first level before hitting the self-reference
        assertTrue(FirebaseMessagingUtil.hasErrorInChain(error, "FIS_AUTH_ERROR"));
    }

    @Test
    public void testHasErrorInChain_selfReferencingCauseNoMatch() {
        SelfReferencingThrowable error = new SelfReferencingThrowable("SOME_OTHER_ERROR");
        // Should terminate without infinite loop despite self-referencing cause
        assertFalse(FirebaseMessagingUtil.hasErrorInChain(error, "FIS_AUTH_ERROR"));
    }

    @Test
    public void testHasErrorInChain_exceedsDepthLimit() {
        int depth = FirebaseMessagingUtil.MAX_CAUSE_CHAIN_DEPTH;
        Throwable deepest = new IOException("FIS_AUTH_ERROR");
        Throwable current = deepest;
        for (int i = 0; i < depth + 1; i++) {
            current = new Exception("wrapper " + i, current);
        }
        assertFalse(FirebaseMessagingUtil.hasErrorInChain(current, "FIS_AUTH_ERROR"));
    }

    @Test
    public void testHasErrorInChain_withinDepthLimit() {
        int depth = FirebaseMessagingUtil.MAX_CAUSE_CHAIN_DEPTH;
        Throwable deepest = new IOException("FIS_AUTH_ERROR");
        Throwable current = deepest;
        for (int i = 0; i < depth - 2; i++) {
            current = new Exception("wrapper " + i, current);
        }
        assertTrue(FirebaseMessagingUtil.hasErrorInChain(current, "FIS_AUTH_ERROR"));
    }

    @Test
    public void testHasErrorInChain_substringMatch() {
        Throwable error = new IOException("Error: FIS_AUTH_ERROR occurred");
        assertTrue(FirebaseMessagingUtil.hasErrorInChain(error, "FIS_AUTH_ERROR"));
    }

    @Test
    public void testHasErrorInChain_emptyErrorMessage() {
        // String.contains("") is always true, so this returns true for any non-null message
        Throwable error = new IOException("anything");
        assertTrue(FirebaseMessagingUtil.hasErrorInChain(error, ""));
    }

    /**
     * A throwable whose getCause() returns itself, simulating a circular reference.
     */
    private static class SelfReferencingThrowable extends Throwable {
        SelfReferencingThrowable(String message) {
            super(message);
        }

        @Override
        public synchronized Throwable getCause() {
            return this;
        }
    }
}

File: app/unit-tests/src/org/commcare/utils/FirebaseMessagingUtilTest.kt
package org.commcare.utils

import org.junit.Assert
import org.junit.Test
import java.io.IOException
import java.util.concurrent.ExecutionException

class FirebaseMessagingUtilTest {
    @Test
    fun testHasErrorInChain_directMatch() {
        val error: Throwable = IOException("FIS_AUTH_ERROR")
        Assert.assertTrue(FirebaseMessagingUtil.hasErrorInChain(error, "FIS_AUTH_ERROR"))
    }

    @Test
    fun testHasErrorInChain_wrappedInExecutionException() {
        val cause = IOException("FIS_AUTH_ERROR")
        val wrapped = ExecutionException(cause)
        Assert.assertTrue(FirebaseMessagingUtil.hasErrorInChain(wrapped, "FIS_AUTH_ERROR"))
    }

    @Test
    fun testHasErrorInChain_serviceNotAvailable() {
        val cause = IOException("SERVICE_NOT_AVAILABLE")
        val wrapped = ExecutionException(cause)
        Assert.assertTrue(FirebaseMessagingUtil.hasErrorInChain(wrapped, "SERVICE_NOT_AVAILABLE"))
    }

    @Test
    fun testHasErrorInChain_noMatch() {
        val error: Throwable = IOException("SOME_OTHER_ERROR")
        Assert.assertFalse(FirebaseMessagingUtil.hasErrorInChain(error, "FIS_AUTH_ERROR"))
    }

    @Test
    fun testHasErrorInChain_nullThrowable() {
        Assert.assertFalse(FirebaseMessagingUtil.hasErrorInChain(null, "FIS_AUTH_ERROR"))
    }

    @Test
    fun testHasErrorInChain_nullMessage() {
        val error = Throwable(null as String?)
        Assert.assertFalse(FirebaseMessagingUtil.hasErrorInChain(error, "FIS_AUTH_ERROR"))
    }

    @Test
    fun testHasErrorInChain_selfReferencingCause() {
        val error = SelfReferencingThrowable("FIS_AUTH_ERROR")
        // Should find the match on the first level before hitting the self-reference
        Assert.assertTrue(FirebaseMessagingUtil.hasErrorInChain(error, "FIS_AUTH_ERROR"))
    }

    @Test
    fun testHasErrorInChain_selfReferencingCauseNoMatch() {
        val error = SelfReferencingThrowable("SOME_OTHER_ERROR")
        // Should terminate without infinite loop despite self-referencing cause
        Assert.assertFalse(FirebaseMessagingUtil.hasErrorInChain(error, "FIS_AUTH_ERROR"))
    }

    @Test
    fun testHasErrorInChain_exceedsDepthLimit() {
        val depth = FirebaseMessagingUtil.MAX_CAUSE_CHAIN_DEPTH
        val deepest: Throwable = IOException("FIS_AUTH_ERROR")
        var current = deepest
        for (i in 0..<depth + 1) {
            current = Exception("wrapper " + i, current)
        }
        Assert.assertFalse(FirebaseMessagingUtil.hasErrorInChain(current, "FIS_AUTH_ERROR"))
    }

    @Test
    fun testHasErrorInChain_withinDepthLimit() {
        val depth = FirebaseMessagingUtil.MAX_CAUSE_CHAIN_DEPTH
        val deepest: Throwable = IOException("FIS_AUTH_ERROR")
        var current = deepest
        for (i in 0..<depth - 2) {
            current = Exception("wrapper " + i, current)
        }
        Assert.assertTrue(FirebaseMessagingUtil.hasErrorInChain(current, "FIS_AUTH_ERROR"))
    }

    @Test
    fun testHasErrorInChain_substringMatch() {
        val error: Throwable = IOException("Error: FIS_AUTH_ERROR occurred")
        Assert.assertTrue(FirebaseMessagingUtil.hasErrorInChain(error, "FIS_AUTH_ERROR"))
    }

    @Test
    fun testHasErrorInChain_emptyErrorMessage() {
        // String.contains("") is always true, so this returns true for any non-null message
        val error: Throwable = IOException("anything")
        Assert.assertTrue(FirebaseMessagingUtil.hasErrorInChain(error, ""))
    }

    /**
     * A throwable whose getCause() returns itself, simulating a circular reference.
     */
    private class SelfReferencingThrowable(
        message: String?,
    ) : Throwable(message) {
        @get:Synchronized
        override val cause: Throwable
            get() = this
    }
}
